### PR TITLE
Add a version catalog plugin

### PR DESCRIPTION
**Pull request by melix**
### Context

This PR adds the concept of a "version catalog" which publishes recommendations (as a TOML file). A version catalog can be imported in `settings.gradle(.kts)` using:

```
dependencyResolutionManagement {
    dependenciesModel("libs") {
        from("my.company:catalog:1.0")
    }
}
```

It is possible to feed a single model from multiple sources:

```
dependencyResolutionManagement {
    dependenciesModel("libs") {
        from("my.company:catalog:1.0")
        from("some.other:catalog:1.0")
    }
}
```
 
in which case, in case of conflict (Gradle issues a warning for example if 2 catalogs use the same aliases), the API lets you include/exclude what to import:

```
dependencyResolutionManagement {
    dependenciesModel("libs") {
        from("my.company:catalog:1.0")
        from("some.other:catalog:1.0") {
            excludeDependency("slf4j")
        }
    }
}
```

### CODE REVIEW

**Comment by melix** [comment]
@bot-gradle test this

**Comment by melix** [comment]
@bot-gradle test this

**Comment by melix** [comment]
@bot-gradle test this

**Comment by melix** [comment]
I'm not convinced by the name "Gradle Platform" yet. It can be a bit confusing. Maybe something like "Gradle recommendation platform". Any other idea is appreciated.

**Comment by melix** [comment]
@bot-gradle test this

**Comment by melix** [comment]
@bot-gradle test this

**Comment by melix** [comment]
@bot-gradle test this

**Comment by melix** [comment]
@bot-gradle test this

**Comment by blindpirate** [comment]
@bot-gradle help

**Review by ljacomet — COMMENTED** [review]
This looks good on the technical side.

I however realised that "Gradle Platform" and its derivative for the name just does not work. It conveys the wrong message.
I find myself liking something like "Dependency Catalog" much more.

File: subprojects/core-api/src/main/java/org/gradle/api/attributes/Usage.java [hunk 1]
```diff
@@ -108,4 +109,13 @@
...
     String SWIFT_API = "swift-api";
+
+    /**
+     * A Gradle platform, packaged as TOML files, for use as recommendations
+     * for dependency and plugin versions.
+     *
+     * @since 6.8
+     */
+    @Incubating
+    String GRADLE_RECOMMENDATIONS = "gradle-recommendations";
```
[the remark this answers is not in the document]
Comment (melix) [reply subprojects/core-api/src/main/java/org/gradle/api/attributes/Usage.java:120]: The pb with "dependency catalog" is that it's not covering everything. In particular plugins and versions refs without libraries.

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/std/DefaultDependenciesModelBuilder.java [hunk 2]
```diff
@@ -105,6 +117,27 @@ private AllDependenciesModel doBuild() {
...
+            return;
+        }
+        DependencyResolutionServices drs = dependencyResolutionServicesSupplier.get();
+        Configuration cnf = drs.getConfigurationContainer().create("incomingPlatformsFor" + StringUtils.capitalize(name));
+        for (Object importedPlatform : importedPlatforms) {
+            Dependency dependency = drs.getDependencyHandler().create(importedPlatform);
+            cnf.getDependencies().add(dependency);
+
+        }
+        cnf.getResolutionStrategy().activateDependencyLocking();
```
[the remark this answers is not in the document]
Comment (melix) [reply subprojects/dependency-management/src/main/java/org/gradle/api/internal/std/DefaultDependenciesModelBuilder.java:131]: Any reason _not_ to?

File: subprojects/core-api/src/main/java/org/gradle/api/attributes/Usage.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (melix) [reply subprojects/core-api/src/main/java/org/gradle/api/attributes/Usage.java:120]: Thinking a bit more, maybe just Gradle catalog is enough: `gradle-catalog`

[the remark this answers is not in the document]
Comment (ljacomet) [reply subprojects/core-api/src/main/java/org/gradle/api/attributes/Usage.java:120]: I am not sure, both plugins and versions refs are about _dependencies_ as well.
I find `gradle` too generic in that context.

**Comment by ljacomet** [comment]
There is something fishy in the number of commits reported in this PR btw.

**Comment by melix** [comment]
> There is something fishy in the number of commits reported in this PR btw.
@ljacomet GitHub was drunk, fixed.

File: subprojects/core-api/src/main/java/org/gradle/api/attributes/Usage.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (melix) [reply subprojects/core-api/src/main/java/org/gradle/api/attributes/Usage.java:120]: Maybe `version-catalog` then? And hope that in the future we don't add more things to it unrelated to versions ;)

**Comment by melix** [comment]
@bot-gradle test this

File: subprojects/core-api/src/main/java/org/gradle/api/attributes/Usage.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (ljacomet) [reply subprojects/core-api/src/main/java/org/gradle/api/attributes/Usage.java]: I like `version-catalog`!

**Comment by melix** [comment]
@bot-gradle test this

**Comment by melix** [comment]
@bot-gradle test this

**Comment by melix** [comment]
@bot-gradle test this

**Comment by melix** [comment]
@bot-gradle test this

**Comment by melix** [comment]
@bot-gradle test this

**Comment by melix** [comment]
@bot-gradle test this

**Comment by melix** [comment]
@bot-gradle test this

**Comment by melix** [comment]
@bot-gradle test this

**Comment by melix** [comment]
@bot-gradle test this

**Comment by melix** [comment]
@bot-gradle test this

**Pull request merged**